### DISCUSSION TITLE
DSL for evrete

### DISCUSSION THREAD

**Discussion by savannahar68** [evrete/evrete#59] [General]
Hey guys, I checked the project, and looks promising.

However, the core philosophy of the rule engine is that developers are no longer dependent on making business changes, and non-tech folks can make these changes via some UI.

For my company, I was thinking of writing a wrapper on top of Evrete, which would have pulled the DSL configured by non-tech folks. But this looks more complex.

I wanted to check if you guys have explored supporting some or one of the standard DSLs(Drools/mvel/jesl/spel) given this might be a very common requirement.

I've already checked [this](https://github.com/evrete/evrete/discussions/36) discussion and found it to be challenging because the evaluate part within execute might check for some conditions and output complex object or boolean. And these condition evaluations could have been simplified if there was a common DSL.

**Comment by andbi** [comment]
You're right about the philosophy, and serving its core purpose was the reason Evrete was created in the first place. I have personally used Drools for over a decade, and IMHO, there's nothing more arduous and error-prone than using mvel in the LHS. With Evrete, tech guys can define arbitrary fields on LHS fact types as functional interfaces and pass them to non-tech guys. Then, for example, an expression like `$obj.field["SOME_KEY"] != 7` turns into `$obj.validKey` w/o changing any POJOs.

The RHS, on the other hand, is a different story. When provided as a string literal, the RHS part is simply compiled into Java byte code. That's the default behavior. For complex actions, MVEL isn't that simpler than plain Java, plus you need to delegate to non-tech folks such important memory tasks as `insert()`, `update()`, and so on. Evrete offers a different approach: tech guys develop a simple syntax for RHS actions and provide a custom implementation of the `LiteralRhsCompiler` SPI interface. They do it once, as simple and domain-specific as possible, and they hide all the memory management calls there. As a result, you get faster code, less end-user support and higher isolation of the DSL language from the system's internals. A total win-win, if you ask me.
